pkg/conversion/mapi2capi/aws_test.go

```diff
@@ -16,43 +16,332 @@ limitations under the License.
-		Expect(capiInfraMachineTemplate).To(Not(BeNil()), "should not have a nil CAPI MachineTemplate")
-	})
+	type awsMAPI2CAPIMachinesetConversionInput struct {
+		machineSetBuilder machinebuilder.MachineSetBuilder
+		infra             *configv1.Infrastructure
+		expectedErrors    []string
+		expectedWarnings  []string
+	}
+
+	var mustConvertAWSProviderSpecToRawExtension = func(spec *mapiv1.AWSMachineProviderConfig) *runtime.RawExtension {
+		if spec == nil {
+			return &runtime.RawExtension{}
+		}
+
+		rawBytes, err := json.Marshal(spec)
+		if err != nil {
+			panic(fmt.Sprintf("unable to convert (marshal) test AWSProviderSpec to runtime.RawExtension: %v", err))
+		}
+
+		return &runtime.RawExtension{
+			Raw: rawBytes,
+		}
+	}
+
+	var _ = DescribeTable("mapi2capi AWS convert MAPI Machine",
+		func(in awsMAPI2CAPIConversionInput) {
+			_, _, warns, err := FromAWSMachineAndInfra(in.machineBuilder.Build(), in.infra).ToMachineAndInfrastructureMachine()
+			Expect(err).To(matchers.ConsistOfMatchErrorSubstrings(in.expectedErrors), "should match expected errors while converting an AWS MAPI Machine to CAPI")
+			Expect(warns).To(matchers.ConsistOfSubstrings(in.expectedWarnings), "should match expected warnings while converting an AWS MAPI Machine to CAPI")
+		},
+
+		// Base Case.
+		Entry("With a Base configuration", awsMAPI2CAPIConversionInput{
+			machineBuilder:   awsMAPIMachineBase,
+			infra:            infra,
+			expectedErrors:   []string{},
+			expectedWarnings: []string{},
+		}),
+
+		// Only Error.
+		Entry("With LoadBalancers", awsMAPI2CAPIConversionInput{
+			machineBuilder: awsMAPIMachineBase.WithProviderSpecBuilder(
+				awsBaseProviderSpec.WithLoadBalancers(
+					[]mapiv1.LoadBalancerReference{{Name: "a", Type: mapiv1.ClassicLoadBalancerType}},
+				),
+			),
+			infra: infra,
+			expectedErrors: []string{
+				fmt.Sprintf("spec.providerSpec.value.loadBalancers: Invalid value: %#v: %s",
+					[]mapiv1.LoadBalancerReference{{Name: "a", Type: mapiv1.ClassicLoadBalancerType}}, "loadBalancers are not supported"),
+			},
+			expectedWarnings: []string{},
+		}),
+		Entry("With DeviceIndex non-zero", awsMAPI2CAPIConversionInput{
+			machineBuilder: awsMAPIMachineBase.WithProviderSpecBuilder(
+				awsBaseProviderSpec.WithDeviceIndex(1),
+			),
+			infra: infra,
+			expectedErrors: []string{
+				fmt.Sprintf("spec.providerSpec.value.deviceIndex: Invalid value: %d: %s", 1, "deviceIndex must be 0 or unset"),
+			},
+			expectedWarnings: []string{},
+		}),
+		Entry("With mismatched region", awsMAPI2CAPIConversionInput{
+			machineBuilder: awsMAPIMachineBase.WithProviderSpecBuilder(
+				awsBaseProviderSpec.WithRegion(awsNonDefaultRegion),
+			),
+			infra: infraWithRegion,
+			expectedErrors: []string{
+				fmt.Sprintf("spec.providerSpec.value.placement.region: Invalid value: %q: %s %q",
+					awsNonDefaultRegion, "placement.region should match infrastructure status value", infraWithRegion.Status.PlatformStatus.AWS.Region),
```

Comment:
Can we avoid `fmt.Sprintf` where possible please, I don't think it's needed here for example